src/iface-modbus.c

```diff
@@ -51,13 +60,116 @@ union modbus_types {
 	uint64_t val_u64;
 };
 
-struct modbus_driver connection_interface;
 struct l_timeout *connect_to;
 struct l_io *modbus_io;
 modbus_t *modbus_ctx;
 iface_modbus_connected_cb_t conn_cb;
 iface_modbus_disconnected_cb_t disconn_cb;
 
+static modbus_t *create_rtu(const char *url)
+{
+	struct serial_rs485 rs485conf;
+	modbus_t *ctx;
+	int mode = MODBUS_RTU_RS232;
+	int fd;
+	int baud_rate;
+	int data_bit;
+	int stop_bit;
+	char parity;
+	char port[256];
+
+	/* Ignoring "serial://" */
+	if (sscanf(&url[8], "%255[^:]:%d , %c , %d , %d", port, &baud_rate,
+		   &parity, &data_bit, &stop_bit) != 5) {
+		l_error("Address (%s) not supported: Invalid format", url);
+		return NULL;
+	}
+
+	l_info("RTU: %s", url);
+
+	fd = open(&url[8], O_RDWR);
+	if (fd < 0)
+		return NULL;
+
+	memset(&rs485conf, 0, sizeof(rs485conf));
+	if (ioctl(fd, TIOCGRS485, &rs485conf) < 0) {
+		mode = MODBUS_RTU_RS232;
+		l_info("Switching to RS-232 ...");
+	} else {
+		mode = MODBUS_RTU_RS485;
+		l_info("Switching to RS-485 ...");
+	}
+
+	close(fd);
+
+	ctx = modbus_new_rtu(port, baud_rate, parity, data_bit, stop_bit);
+
+	if (!ctx)
+		return NULL;
+
+	modbus_rtu_set_serial_mode(ctx, mode);
+	modbus_rtu_set_rts(ctx, MODBUS_RTU_RTS_NONE);
+
+	return ctx;
+}
+
+static modbus_t *create_tcp(const char *url)
+{
+	char hostname[128];
+	char port[8];
+
+	memset(hostname, 0, sizeof(hostname));
+	memset(port, 0, sizeof(port));
+
+	/* Ignoring "tcp://" */
+	if (sscanf(&url[6], "%127[^:]:%7s", hostname, port) != 2) {
+		l_error("Address (%s) not supported: Invalid format", url);
+		return NULL;
+	}
+
+	l_info("TCP: %s", url);
+
+	return modbus_new_tcp_pi(hostname, port);
+}
+
+static void destroy(modbus_t *ctx)
+{
+	modbus_close(ctx);
+	modbus_free(ctx);
+}
+
+static int read_bool(modbus_t *ctx, uint16_t addr, bool *out)
+{
+	uint8_t val_u8 = 0;
+	int ret;
+
+	ret = modbus_read_input_bits(ctx, addr, 1, &val_u8);
+	if (ret > 0)
+		*out = (val_u8 ? true : false);
+
+	return ret;
+}
+
+static int read_byte(modbus_t *ctx, uint16_t addr, uint8_t *out)
```

Comment:
I do not think the read_* functions are necessary anymore since they are acting as a wrapper.

Reply:
genius. done